engine/packages/pegboard-runner/src/lib.rs

```diff
@@ -79,52 +86,142 @@ impl CustomServeTrait for PegboardRunnerWsCustomServe {
 			.await
 			.context("failed to initialize runner connection")?;
 
-		// Subscribe to pubsub topic for this runner before accepting the client websocket so
-		// that failures can be retried by the proxy.
+		// Subscribe before accepting the client websocket so that failures can be retried by the proxy.
 		let topic =
 			pegboard::pubsub_subjects::RunnerReceiverSubject::new(conn.runner_id).to_string();
-		tracing::debug!(%topic, "subscribing to runner receiver topic");
+		let eviction_topic =
+			pegboard::pubsub_subjects::RunnerEvictionByIdSubject::new(conn.runner_id).to_string();
+		let eviction_topic2 = pegboard::pubsub_subjects::RunnerEvictionByNameSubject::new(
+			conn.namespace_id,
+			&conn.runner_name,
+			&conn.runner_key,
+		)
+		.to_string();
+
+		tracing::debug!(%topic, %eviction_topic, %eviction_topic2, "subscribing to runner topics");
 		let sub = ups
 			.subscribe(&topic)
 			.await
 			.with_context(|| format!("failed to subscribe to runner receiver topic: {}", topic))?;
+		let mut eviction_sub = ups.subscribe(&eviction_topic).await.with_context(|| {
+			format!(
+				"failed to subscribe to runner eviction topic: {}",
+				eviction_topic
+			)
+		})?;
+		let mut eviction_sub2 = ups.subscribe(&eviction_topic2).await.with_context(|| {
+			format!(
+				"failed to subscribe to runner eviction topic: {}",
+				eviction_topic2
+			)
+		})?;
+
+		// Publish eviction message to evict any currently connected runners with the same id or ns id +
+		// runner name + runner key. This happens after subscribing to prevent race conditions.
+		tokio::try_join!(
+			async {
+				ups.publish(&eviction_topic, &[], PublishOpts::broadcast())
+					.await?;
+				// Because we will receive our own message, skip the first message in the sub
+				eviction_sub.next().await
+			},
+			async {
+				ups.publish(&eviction_topic2, &[], PublishOpts::broadcast())
+					.await?;
+				eviction_sub2.next().await
+			},
+		)?;
 
-		// Forward pubsub -> WebSocket
-		let mut pubsub_to_client = tokio::spawn(pubsub_to_client_task::task(conn.clone(), sub));
+		let (tunnel_to_ws_abort_tx, tunnel_to_ws_abort_rx) = watch::channel(());
+		let (ws_to_tunnel_abort_tx, ws_to_tunnel_abort_rx) = watch::channel(());
+		let (ping_abort_tx, ping_abort_rx) = watch::channel(());
+
+		let tunnel_to_ws = tokio::spawn(tunnel_to_ws_task::task(
```

Comment:
imo this naming is confusing. most of this file is tunnel logic, but it's also where all of the other ws<->pb forwarding logic happens.